### PR TITLE
[dagster_model] support default values

### PR DESCRIPTION
**Pull request by alangenfeld**
* support empty models
* support setting default values in the model declaration
  * ensure things like [] and {} are not shared

## How I Tested These Changes

added tests

### CODE REVIEW

**Comment by alangenfeld** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22374?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#22374** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22374?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#22366** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22366?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @alangenfeld and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/dagster/dagster/_model/decorator.py [hunk 1]
```diff
@@ -260,13 +287,16 @@ def __call__(self, cls, **kwargs):
...
             self._build_checked_new_str(),
-            _GENERATED_NEW,
+            _CHECKED_NEW,
         )
 
         return self._nt_base.__new__(cls, **kwargs)
 
     def _build_checked_new_str(self) -> str:
-        kw_args = ", ".join(self._field_set.keys())
+        # field_set = getattr(self._cls, "__annotations__", {})
```
Comment (schrockn) [inline python_modules/dagster/dagster/_model/decorator.py:296]: rm?

File: python_modules/dagster/dagster/_model/decorator.py [hunk 2]
```diff
@@ -41,21 +44,42 @@ def _namedtuple_model_transform(
...
     """
-    base = NamedTuple(f"_{cls.__name__}", cls.__annotations__.items())
+    field_set = getattr(cls, "__annotations__", {})
+    defaults = {name: getattr(cls, name) for name in field_set.keys() if hasattr(cls, name)}
+
+    base = NamedTuple(f"_{cls.__name__}", field_set.items())
     nt_new = base.__new__
     if checked:
+        eval_ctx = EvalContext.capture_from_frame(1 + decorator_frames)
+        eval_ctx.local_ns[_DEFAULT_VALUES_KEY] = defaults
```
Comment (schrockn) [inline python_modules/dagster/dagster/_model/decorator.py:54]: shouldn't this mutation of `local_ns` happen in the constructor?

File: python_modules/dagster/dagster/_model/decorator.py [hunk 3]
```diff
@@ -41,21 +44,42 @@ def _namedtuple_model_transform(
...
             base,
+            eval_ctx,
+            1 if with_new else 0,
         )
         base.__new__ = jit_checked_new  # type: ignore
 
+    elif defaults:
+        # allow arbitrary ordering of default values since we are kwarg only by generating a __new__ impl
+        eval_ctx = EvalContext({}, {})
+        eval_ctx.local_ns[_DEFAULT_VALUES_KEY] = defaults
```
Comment (schrockn) [inline python_modules/dagster/dagster/_model/decorator.py:67]: same here. not a fan of mutation outside ctor

File: python_modules/dagster/dagster/_model/decorator.py [hunk 4]
```diff
@@ -41,21 +44,42 @@ def _namedtuple_model_transform(
...
+            defaults,
             base,
+            eval_ctx,
+            1 if with_new else 0,
         )
         base.__new__ = jit_checked_new  # type: ignore
 
+    elif defaults:
+        # allow arbitrary ordering of default values since we are kwarg only by generating a __new__ impl
+        eval_ctx = EvalContext({}, {})
```
Comment (schrockn) [inline python_modules/dagster/dagster/_model/decorator.py:66]: kwargs please for self-documentation

File: python_modules/dagster/dagster/_model/decorator.py [hunk 5]
```diff
@@ -66,6 +90,7 @@ def _namedtuple_model_transform(
             _MODEL_MARKER_FIELD: _MODEL_MARKER_VALUE,
             "__annotations__": cls.__annotations__,
             "__nt_new__": nt_new,
+            "__bool__": _true,
```
Comment (schrockn) [inline python_modules/dagster/dagster/_model/decorator.py:93]: ❤️

File: python_modules/dagster/dagster/_model/decorator.py [hunk 6]
```diff
@@ -276,18 +306,72 @@ def _build_checked_new_str(self) -> str:
...
+
+    kw_args_str = ""
+    if kw_args:
+        kw_args_str = f", *, {', '.join(kw_args)}"
+
+    set_calls_str = ""
+    if set_calls:
+        set_calls_str = "\n    ".join(set_calls)
+
+    return kw_args_str, set_calls_str
```
Comment (schrockn) [inline python_modules/dagster/dagster/_model/decorator.py:365]: there are getting complicated enough where some unit tests could be good, if nothing else than for documentation purposes.

**Review by schrockn — CHANGES_REQUESTED** [review]
excited for this to go in. enough feedback to do a spin

File: python_modules/dagster/dagster/_check/__init__.py [hunk 7]
```diff
@@ -1849,13 +1849,23 @@ class EvalContext(NamedTuple):
...
+
+        if add_to_local_ns:
+            local_ns.update(add_to_local_ns)
+
         return EvalContext(
-            # copy to not mess up frame data
-            ctx_frame.f_globals.copy(),
-            ctx_frame.f_locals.copy(),
+            global_ns=global_ns,
+            local_ns=local_ns,
```
Comment (schrockn) [inline python_modules/dagster/dagster/_check/__init__.py:1868]: at this point this is comment-worthy

File: python_modules/dagster/dagster/_model/decorator.py [hunk 8]
```diff
@@ -41,21 +44,43 @@ def _namedtuple_model_transform(
...
-    base = NamedTuple(f"_{cls.__name__}", cls.__annotations__.items())
+    field_set = getattr(cls, "__annotations__", {})
+    defaults = {name: getattr(cls, name) for name in field_set.keys() if hasattr(cls, name)}
+
+    base = NamedTuple(f"_{cls.__name__}", field_set.items())
     nt_new = base.__new__
     if checked:
+        eval_ctx = EvalContext.capture_from_frame(
+            1 + decorator_frames,
+            add_to_local_ns={_DEFAULT_VALUES_KEY: defaults},
```
Comment (schrockn) [inline python_modules/dagster/dagster/_model/decorator.py:55]: Yeah this is the starting to get to "magical incantations" territory here with the hardcoded keys and the like.

**Review by schrockn — CHANGES_REQUESTED** [review]
Can you add comments to the places where I commented? This is getting fairly obtuse at this point.

**Review by schrockn — APPROVED** [review]

**Pull request merged**